src/color-extractor.ts

```diff
@@ -0,0 +1,417 @@
+ * can get dominant colors of the image by using getDominantColors. Dominant 
+ * colors are found by running k means on the colors in the image. Dominant 
+ * colors that are too close to certain specified colors are filtered out 
+ * before the dominant colors are returned.
+ */
+export class DominantColorComputer {
+  private colorFilter: ColorFilter;
+  private kMeansCalculator: KMeansCalculator
+
+  constructor(data: ImageData) {
+    const [dataMap, keyArray, pixelCount] = this.getDataMap(data);
+    this.colorFilter = new ColorFilter();
+    this.kMeansCalculator = new KMeansCalculator(dataMap, keyArray, pixelCount);
+  }
+
+  /**
+   * Takes in an ImageData object and finds the k most dominant colors from
+   * the image data. Also returns the number of pixels assoiated with each 
+   * color's cluster. Dominant colors that are too close to colors in 
+   * colorsToIgnore are ignored.
+   * 
+   * @param numClusters     Number of dominant colors to find
+   * @param colorsToIgnore  Array of colors as rgb strings to ignore
+   *                        Format of rgb strings is [red][green][blue],
+   *                        where [red], [green], and [blue] are the values
+   *                        of the red, green, and blue channels represented as
+   *                        strings. Values range from 0 to 255 and are padded
+   *                        with zeros in the front so each channel value has
+   *                        length 3, and the rgb string has length 9.
+   * @param threshold       Threshold for similarity in LAB color space
+   *                        for ignoring dominant color
+   */
+  getDominantColors(numClusters: number, colorsToIgnore: Array<string>,
+    threshold: number): Array<any> {
+    let centers = new Array<string>();
+    let centerCounts, centerToColor;
+    while (centers.length === 0) {
+      [centers, centerCounts, centerToColor] = this.kMeansCalculator.kMeans(numClusters);
+      console.log(centers);
+      [centers, centerCounts] = this.colorFilter.filterColors(centers, centerCounts, centerToColor, 
+        colorsToIgnore, threshold);
+      console.log(centers);
+    }
+
+    return [centers, centerCounts];
+  }
+
+  /**
+   * Converts an ImageData object into a map from colors to counts in the 
+   * image. Also returns an array of all colors and the total number of pixels
+   * in the image. 
+   */
+  private getDataMap(myData: ImageData): Array<any> {
+    let dataMap = new Map<string, number>();
+    let keyArray = new Array<string>();
+    let total = 0;
+    const numColorChannels = 4;
+
+    for (let i = 0; i < myData.data.length; i += numColorChannels) {
+      if (myData.data[i + 3] > 0) {
+        const r = this.padColorString(myData.data[i].toString());
+        const g = this.padColorString(myData.data[i + 1].toString());
+        const b = this.padColorString(myData.data[i + 2].toString());
+
+        const key = r + g + b;
```

Comment:
So the keys are strings, but I noticed later we need to convert them back to numerical values.  Would it make sense to use number keys instead?  E.g.

  key = myData.data[i] | (myData.data[i + 1] << 8) | (myData.data[i + 2] << 16);

Then we can extract the channel values using bitwise ops later, and avoid string parsing/manipulation?